jerry-core/ecma/operations/ecma-function-object.c

```diff
@@ -874,13 +762,117 @@ ecma_op_function_construct (ecma_object_t *func_obj_p, /**< Function object */
 
     /* 5. */
     ret_value = ecma_op_function_construct (target_func_obj_p,
+                                            this_arg_value,
                                             merged_args_list_p,
                                             merged_args_list_len);
 
     JMEM_FINALIZE_LOCAL_ARRAY (merged_args_list_p);
 
     return ret_value;
   }
+
+  ecma_object_type_t type = ecma_get_object_type (func_obj_p);
+
+#ifndef CONFIG_DISABLE_ES2015_ARROW_FUNCTION
+  if (JERRY_UNLIKELY (type == ECMA_OBJECT_TYPE_ARROW_FUNCTION))
+  {
+    return ecma_raise_type_error (ECMA_ERR_MSG ("Arrow functions have no constructor."));
+  }
+#endif /* CONFIG_DISABLE_ES2015_ARROW_FUNCTION */
+
+  if (JERRY_UNLIKELY (type == ECMA_OBJECT_TYPE_FUNCTION && ecma_get_object_is_builtin (func_obj_p)))
+  {
+    if (ecma_builtin_function_is_routine (func_obj_p))
+    {
+      return ecma_raise_type_error (ECMA_ERR_MSG ("Built-in routines have no constructor."));
+    }
+
+    return ecma_builtin_dispatch_construct (func_obj_p,
+                                            arguments_list_p,
+                                            arguments_list_len);
+  }
+
+  ecma_object_t *new_this_obj_p = NULL;
+
+  if (JERRY_LIKELY (this_arg_value == ECMA_VALUE_UNDEFINED))
+  {
+    /* 5. */
+    ecma_value_t prototype_prop_value = ecma_op_object_get_by_magic_id (func_obj_p,
+                                                                        LIT_MAGIC_STRING_PROTOTYPE);
+
+    if (ECMA_IS_VALUE_ERROR (prototype_prop_value))
+    {
+      return prototype_prop_value;
+    }
+
+    /* 1., 2., 4. */
+    if (ecma_is_value_object (prototype_prop_value))
+    {
+      /* 6. */
+      new_this_obj_p = ecma_create_object (ecma_get_object_from_value (prototype_prop_value),
+                                           0,
+                                           ECMA_OBJECT_TYPE_GENERAL);
+    }
+    else
+    {
+      /* 7. */
+      ecma_object_t *prototype_p = ecma_builtin_get (ECMA_BUILTIN_ID_OBJECT_PROTOTYPE);
+
+      new_this_obj_p = ecma_create_object (prototype_p, 0, ECMA_OBJECT_TYPE_GENERAL);
+
+      ecma_deref_object (prototype_p);
+    }
+
+    ecma_free_value (prototype_prop_value);
+
+    this_arg_value = ecma_make_object_value (new_this_obj_p);
+  }
+
+  /* 8. */
+  ecma_value_t ret_value;
+
+  switch (type)
```

Comment:
An `if-else` statement fits better here than a `switch`

Reply:
Currently it fits better but for further [cases](https://github.com/jerryscript-project/jerryscript/pull/2439/files#diff-c790ca6e2b99859fd40e7cce2885c2fbR1068) switch is more suitable.

Reply:
OK, leave as it is.